### PR TITLE
Added two new challenges class test to improve test coverage

### PR DESCRIPTION
**Pull request by samymessouak**
# Add Comprehensive UI Tests for NewChallengeScreen and CreateAChallengeScreen

## Acknowledgment
* After the valuable feedback from my coaches, I realized the importance of combining feature and test code in a single PR. This helps maintain clarity, context, and better review practices.
* Moving forward, I will ensure not to push tests alone without their associated features.

## Overview
This pull request introduces comprehensive UI and functionality tests for the **NewChallengeScreen** and **CreateAChallengeScreen**. These tests validate the robustness of user interactions and ensure seamless integration with Firestore.

### 1. NewChallengeScreen Tests
This task focused on validating the functionality and design of the revamped **NewChallengeScreen**, ensuring a seamless user experience and robust Firestore integration. The tests included:

- **Challenge Deletion**: Unit tests to verify the successful deletion of challenges directly from Firestore via the **NewChallengeScreen**, ensuring data integrity and synchronization.
- **UI Element Testing**: Detailed tests for all newly added UI components to confirm their proper functioning and visual consistency with the application's design standards.
- **End-to-End Testing**: Verification of the overall interaction flow, ensuring users can effectively navigate, view, and manage challenges through the updated screen.

These tests ensure that the **NewChallengeScreen** operates reliably, providing an enhanced challenge management experience while maintaining the integrity of the Firestore backend.

### 2. CreateAChallengeScreen Tests
This task involved designing and implementing comprehensive tests to ensure the **CreateAChallengeScreen** meets all functional requirements for user interaction and Firestore integration. The aspects covered included:

- **Challenge Creation**: Verification of challenges successfully being created and stored in the Firestore database, ensuring correct data handling and synchronization.
- **Button Functionality**: Testing all button interactions to confirm they perform their intended actions, such as creating challenges and navigating the user interface.
- **UI Validation**: Thorough checks to ensure that the screen is displayed correctly, with all components properly aligned and styled in accordance with design specifications.

These tests ensure that the **CreateAChallengeScreen** provides a seamless and reliable user experience while maintaining both functional and aesthetic consistency with the application's overall design.

## Summary of Changes
- Added unit and UI tests for **NewChallengeScreen**.
- Added comprehensive interaction tests for **CreateAChallengeScreen**.
- Verified Firestore integration for challenge creation and deletion.

These tests aim to improve the quality, reliability, and user experience of both screens.

### CODE REVIEW

**Review by NBCCF025 — APPROVED** [review]
This PR bring usefull tests for `CreateAChallengeScreen.kt` and `NewChallengeScreen.kt`.
It improve test coverage and ensure the screens have the proper behaviour.
I approve this PR, good job.

image

File: app/src/androidTest/java/com/github/se/signify/ui/screens/challenge/CreateAChallengeScreenTest.kt
Comment (ziyadbagh) [inline app/src/androidTest/java/com/github/se/signify/ui/screens/challenge/CreateAChallengeScreenTest.kt:1]: 1. **Structure and Initialization**: Well-organized with comprehensive setup. Mocks and `doAnswer` are effectively used to control test conditions.

2. **Composable Setup**: Proper use of `composeTestRule.setContent` to render `CreateAChallengeScreen` and trigger `userViewModel.getFriendsList()`.

3. **Tests**: Clear, well-named tests with precise assertions (`assertIsDisplayed()`, `assertIsEnabled()`). Good coverage of core UI and interaction cases.

4. **Mock Verifications**: `verify` calls on `sendChallengeRequest` and `addOngoingChallenge` are correctly used to confirm repository interactions.

5. **Edge Cases**: Handled well, like empty friend lists and dialog canceling. Consider adding error case tests for methods like `getFriendsList`.

6. **Enhancements**:
   - Add verifications for UI state transitions (e.g., `SendChallengeButton` enabled after selecting a mode).
   - Parameterize `friends` for varied scenarios (e.g., more friends, duplicates).
   - Test alternate challenge modes if available.

7. **Comments**: Useful comments; consider Javadoc-style summaries for each test case.

8. **Coverage**: Good, but consider tests for dialog content, friend list boundaries, and all available challenge modes.

Overall, solid coverage and clear testing.

File: app/src/androidTest/java/com/github/se/signify/ui/screens/challenge/NewChallengeScreenTest.kt
Comment (ziyadbagh) [inline app/src/androidTest/java/com/github/se/signify/ui/screens/challenge/NewChallengeScreenTest.kt:1]: This test suite is similarly well-structured, following the previous example. It uses effective setup with mocks (`doAnswer` for `getOngoingChallenges` and `removeOngoingChallenge`) and covers essential UI elements and actions, like button clicks and list displays. Assertions and `verify` calls on navigation and repository methods are well-placed, and edge cases like deletion are handled with appropriate state checks. As with the previous code, consider minor enhancements, like error case testing and brief inline comments for clarity. Overall, it aligns well with the previous structure, providing comprehensive coverage.

**Review by ziyadbagh — APPROVED** [review]
As you mentioned, including both feature and test code in the same PR will indeed help maintain clarity and streamline the review process.

Good job !!

**Pull request merged**